velox/connectors/hive/iceberg/IcebergPartitionIdGenerator.cpp

```diff
@@ -0,0 +1,202 @@
+    key->resize(maxPartitions);
+  }
+}
+
+void IcebergPartitionIdGenerator::savePartitionValues(
+    uint64_t partitionId,
+    const RowVectorPtr& input,
+    vector_size_t row) {
+  for (auto i = 0; i < partitionChannels_.size(); ++i) {
+    partitionValues_->childAt(i)->copy(
+        input->childAt(i).get(), partitionId, row, 1);
+  }
+}
+
+void IcebergPartitionIdGenerator::run(
+    const RowVectorPtr& input,
+    raw_vector<uint64_t>& result) {
+  const auto numRows = input->size();
+  result.resize(numRows);
+  std::vector<VectorPtr> columns;
+  std::vector<std::string> names;
+  std::vector<TypePtr> types;
+  const int32_t transformCount = columnTransforms_.size();
+  columns.reserve(transformCount);
+  names.reserve(transformCount);
+  types.reserve(transformCount);
+  for (const auto& columnTransform : columnTransforms_) {
+    names.emplace_back(columnTransform.columnName());
+    types.emplace_back(columnTransform.resultType());
+    columns.emplace_back(columnTransform.transform(input));
+  }
+  const auto rowVector = std::make_shared<RowVector>(
+      pool_,
+      ROW(std::move(names), std::move(types)),
+      nullptr,
+      numRows,
+      columns);
+
+  // Compute value IDs using VectorHashers and store these in 'result'.
+  computeValueIds(rowVector, result);
+
+  // Convert value IDs in 'result' into partition IDs using partitionIds
+  // mapping. Update 'result' in place.
+  for (auto i = 0; i < numRows; ++i) {
+    auto valueId = result[i];
+    if (auto it = partitionIds_.find(valueId); it != partitionIds_.end()) {
+      result[i] = it->second;
+    } else {
+      uint64_t nextPartitionId = partitionIds_.size();
+      VELOX_USER_CHECK_LT(
+          nextPartitionId,
+          maxPartitions_,
+          "Exceeded limit of {} distinct partitions.",
+          maxPartitions_);
+
+      partitionIds_.emplace(valueId, nextPartitionId);
+      savePartitionValues(nextPartitionId, rowVector, i);
+      result[i] = nextPartitionId;
+    }
+  }
+}
+
+std::vector<std::pair<std::string, std::string>>
+IcebergPartitionIdGenerator::extractPartitionKeyValues(
+    const RowVectorPtr& partitionsVector,
+    vector_size_t row,
+    const std::string& nullValueString) const {
+  std::vector<std::pair<std::string, std::string>> partitionKeyValues;
+  VELOX_DCHECK_EQ(
+      partitionsVector->childrenSize(),
+      columnTransforms_.size(),
+      "Partition values and partition transform doe not match.");
+  for (auto i = 0; i < partitionsVector->childrenSize(); i++) {
+    partitionKeyValues.push_back(VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+        makePartitionKeyValueString,
+        partitionsVector->childAt(i)->typeKind(),
+        partitionsVector->childAt(i)->loadedVector(),
+        row,
+        asRowType(partitionsVector->type())->nameOf(i),
+        columnTransforms_[i],
+        nullValueString));
+  }
+  return partitionKeyValues;
+}
+
+std::string IcebergPartitionIdGenerator::partitionName(
```

Comment:
Doo you need to add `nullValueName` to partitionName? The scope has been in Iceberg, makePartitionKeyValueString can use "null" directly.